### PR TITLE
sqltostr: fix str_to_date('2018-6-14', '%Y-%m-%d') bug

### LINKED ISSUE #6835: Behavior of str_to_date() is different from mysql
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
`select str_to_date('2018-6-14', '%Y-%m-%d');`
`select str_to_date('2018-6-14', '%Y-%c-%d');`

2. What did you expect to see?
'2018-06-14'

3. What did you see instead?
NULL

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v2.1.0-alpha-48-ge2c8c4e
UTC Build Time:  2018-06-13 12:05:19

### PR DESCRIPTION
**Pull request by mail2fish**
## What have you changed? (mandatory)

fix #6835

## What are the type of the changes (mandatory)?

bug fixed

## How has this PR been tested (mandatory)?

make dev

## Refer to a related PR or issue link (optional)

https://github.com/pingcap/tidb/issues/6835

### CODE REVIEW

**Comment by lysu** [comment]
@mail2fish Thanks for your PR!

**Comment by zz-jason** [comment]
Review status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @mail2fish)

---

*[types/time.go, line 2321 at r1](https://reviewable.io/reviews/pingcap/tidb/6919#-LG-DbJJBDae9AmxRJGA:-LG-DbJJBDae9AmxRJGB:bu31a0v) ([raw file](https://github.com/pingcap/tidb/blob/cf37e046fb0ede745ef480f5b5f12325d27ce978/types/time.go#L2321)):*
> ```Go
> }
> 
> var digitRegex = regexp.MustCompile("^[0-9]+")
> ```

It'e better to put these two regex patterns together to a code block and make some comment for each one.

---

*[types/time.go, line 2348 at r1](https://reviewable.io/reviews/pingcap/tidb/6919#-LG-EYfX14z4gdodDfKD:-LG-EYfX14z4gdodDfKE:b-xaw42x) ([raw file](https://github.com/pingcap/tidb/blob/cf37e046fb0ede745ef480f5b5f12325d27ce978/types/time.go#L2348)):*
> ```Go
> 
> 	if length > 2 {
> 		result = result[2:]
> ```

Since `result` is not used in the following execution, we can remove this line.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pingcap/tidb/6919)*

File: types/time.go [hunk 1]
```diff
@@ -2349,12 +2341,21 @@ func parseTwoNumeric(input string) (int, string) {
 }
 
 func dayOfMonthNumeric(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
-	v, remain := parseTwoNumeric(input) // 0..31
-	if len(remain) == len(input) || v > 31 {
+	result := digitRegex.FindString(input) // 0..31
+	length := len(result)
+
+	if length > 2 {
```
Comment (zhexuany) [inline types/time.go]: is it better to add some comments here why do you only take the suffix after index 2?

**Comment by zz-jason** [comment]
Review status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @mail2fish and @zhexuany)

---

*[types/time.go, line 2357 at r2](https://reviewable.io/reviews/pingcap/tidb/6919#-LG-TLqsD3s-NVJLGt8S:-LG-TLqsD3s-NVJLGt8T:bvaj5eg) ([raw file](https://github.com/pingcap/tidb/blob/5bc75d48c5df29544100430be4756ef2bf4d0f3d/types/time.go#L2357)):*
> ```Go
> 	v, ok := parseDigits(input, length)
> 
> 	if !ok || v >= 32 {
> ```

I think we should keep the origin `v > 31`, this check form is consistent with the validation check of month, which is `v > 12` at line 2458

---

*[types/time.go, line 2447 at r2](https://reviewable.io/reviews/pingcap/tidb/6919#-LG-T33x9xbCVaWvVDMI:-LG-T33x9xbCVaWvVDMJ:bm78oid) ([raw file](https://github.com/pingcap/tidb/blob/5bc75d48c5df29544100430be4756ef2bf4d0f3d/types/time.go#L2447)):*
> ```Go
> 
> func monthNumeric(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
> ```

can we remove this empty line?

---


*Comments from [Reviewable](https://reviewable.io/reviews/pingcap/tidb/6919)*

File: types/time.go [hunk 2]
```diff
@@ -2108,26 +2108,26 @@ var monthAbbrev = map[string]gotime.Month{
...
-	"%m": monthNumericTwoDigits,      // Month, numeric (00..12)
-	"%p": isAMOrPM,                   // AM or PM
-	"%r": time12Hour,                 // Time, 12-hour (hh:mm:ss followed by AM or PM)
-	"%s": secondsNumeric,             // Seconds (00..59)
-	"%S": secondsNumeric,             // Seconds (00..59)
-	"%T": time24Hour,                 // Time, 24-hour (hh:mm:ss)
-	"%Y": yearNumericFourDigits,      // Year, numeric, four digits
+	"%b": abbreviatedMonth,      // Abbreviated month name (Jan..Dec)
+	"%c": monthNumeric,          // Month, numeric (0..12)
+	"%d": dayOfMonthNumeric,     // Day of the month, numeric (00..31)
```
Comment (shenli) [inline types/time.go]: Please update the comment.

File: types/time.go [hunk 3]
```diff
@@ -2108,26 +2108,26 @@ var monthAbbrev = map[string]gotime.Month{
...
+	"%f": microSeconds,          // Microseconds (000000..999999)
+	"%h": hour24TwoDigits,       // Hour (01..12)
+	"%H": hour24TwoDigits,       // Hour (01..12)
+	"%I": hour24TwoDigits,       // Hour (01..12)
+	"%i": minutesNumeric,        // Minutes, numeric (00..59)
+	"%j": dayOfYearThreeDigits,  // Day of year (001..366)
+	"%k": hour24Numeric,         // Hour (0..23)
+	"%l": hour12Numeric,         // Hour (1..12)
+	"%M": fullNameMonth,         // Month name (January..December)
+	"%m": monthNumeric,          // Month, numeric (00..12)
```
Comment (shenli) [inline types/time.go]: ditto

File: types/time.go (hunk 2 above)
Comment (mail2fish) [reply types/time.go]: Which the comment be needed to update?

Comment (zz-jason) [reply types/time.go]: update "(00..31)" to "(0..31)"

File: types/time.go [hunk 4]
```diff
@@ -2450,12 +2444,22 @@ func fullNameMonth(t *MysqlTime, input string, ctx map[string]int) (string, bool
 }
 
 func monthNumeric(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
-	v, rem := parseTwoNumeric(input)
-	if len(rem) == len(input) || v > 12 {
-		return rem, false
+
```
Comment (zz-jason) [inline types/time.go]: please remove this empty line.

**Comment by zz-jason** [comment]
@mail2fish Please remove the empty line between the function declaration and the first code of that function, that empty line is meaningless and redundant.

File: types/time.go (hunk 3 above)
Comment (shenli) [reply types/time.go]: Please update the comment.

File: types/time.go [hunk 5]
```diff
@@ -2318,15 +2318,10 @@ func isAMOrPM(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
...
-func dayOfMonthNumericTwoDigits(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
-	v, succ := parseDigits(input, 2)
-	if !succ || v >= 32 {
-		return input, false
-	}
-	t.day = uint8(v)
-	return input[2:], true
-}
+// digitRegex: it was used to scan a variable-length monthly day or month in the string. Ex:  "01" or "1" or "30"
+var digitRegex = regexp.MustCompile("^[0-9]+")
```
Comment (shenli) [inline types/time.go]: `^[0-9]{1,2}` is suitable?

File: types/time.go [hunk 6]
```diff
@@ -2349,32 +2344,58 @@ func parseTwoNumeric(input string) (int, string) {
...
-	if len(remain) == len(input) || v > 31 {
+	result := digitRegex.FindString(input) // 0..31
+	length := len(result)
+
+	// Some time the input is a consecutive digital string. Ex: 20181120
+	if length > 2 {
+		length = 2
+	}
+
+	v, ok := parseDigits(input, length)
```
Comment (shenli) [inline types/time.go]: For  "20181120" the v would be 20?

File: types/time.go [hunk 7]
```diff
@@ -2349,32 +2344,58 @@ func parseTwoNumeric(input string) (int, string) {
...
+
+	if !ok || v > 31 {
 		return input, false
 	}
-	t.month = uint8(v)
-	return remain, true
+	t.day = uint8(v)
```
Comment (shenli) [inline types/time.go]: `t.month = uint8(v)` is a bug?

File: types/time.go (hunk 5 above)
Comment (mail2fish) [reply types/time.go]: Yes, It's better

File: types/time.go (hunk 7 above)
Comment (mail2fish) [reply types/time.go]: Yes, I think so.

File: types/time.go [hunk 8]
```diff
@@ -2450,12 +2461,22 @@ func fullNameMonth(t *MysqlTime, input string, ctx map[string]int) (string, bool
...
 
 func monthNumeric(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
-	v, rem := parseTwoNumeric(input)
-	if len(rem) == len(input) || v > 12 {
-		return rem, false
+
+	result := digitRegex.FindString(input) // 1..12
+	length := len(result)
+
+	// Some time the input is a consecutive digital string. Ex: 20181120
```
Comment (zhexuany) [inline types/time.go]: You want to say `e.g.`?

Comment (mail2fish) [reply types/time.go]: Just forgot to remove those lines

**Comment by shenli** [comment]
LGTM

File: types/time.go [hunk 9]
```diff
@@ -2108,26 +2108,26 @@ var monthAbbrev = map[string]gotime.Month{
...
-	"%M": fullNameMonth,              // Month name (January..December)
-	"%m": monthNumericTwoDigits,      // Month, numeric (00..12)
-	"%p": isAMOrPM,                   // AM or PM
-	"%r": time12Hour,                 // Time, 12-hour (hh:mm:ss followed by AM or PM)
-	"%s": secondsNumeric,             // Seconds (00..59)
-	"%S": secondsNumeric,             // Seconds (00..59)
-	"%T": time24Hour,                 // Time, 24-hour (hh:mm:ss)
-	"%Y": yearNumericFourDigits,      // Year, numeric, four digits
+	"%b": abbreviatedMonth,      // Abbreviated month name (Jan..Dec)
+	"%c": monthNumeric,          // Month, numeric (0..12)
```
Comment (winoros) [inline types/time.go]: These `0` should be `00`?

Comment (mail2fish) [reply types/time.go]: from the https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-format

%c Month, numeric (0..12)

**Pull request merged**